Support multiple DNS queries

davydog187 opened the pull request:
This PR extends the API to allow for a list of DNS queries. This is useful if you have multiple services that you want to cluster together

davydog187 committed:
Support multiple DNS queries

This PR extends the API to allow for a list of DNS queries. This is
useful if you have multiple services that you want to cluster together

davydog187 committed:
fix example

davydog187 commented on lib/dns_cluster.ex:
```diff
@@ -90,7 +91,8 @@ defmodule DNSCluster do
       {:ok, other} ->
-        raise ArgumentError, "expected :query to be a string, got: #{inspect(other)}"
+        raise ArgumentError,
+              "expected :query to be a string or list of strings, got: #{inspect(other)}"
```
note that we don't validate that the list contains only strings. Happy to add that if we think its necessary (seems like its not to me)

davydog187 committed:
expand the tests

davydog187 commented:
This PR needs a little more thought due to the basename restriction on connected nodes. I'll try to give this more time tomorrow

joshk commented:
This is really awesome! I was going to create a similar PR but was happy to see it was already done!

Would it be possible to accept a list of domains separated by commas? 
eg. `"myapp.internal"` or `["foo.internal", "bar.internal"]` or `"foo.internal,bar.internal"`

This would make it a lot easier to configure via environment variables. 
eg.`DNS_CLUSTER_QUERY="foo.internal,bar.internal"`

davydog187 commented:
hello @joshk! I still need to make some tweaks to this PR before it can be merged, unfortunately.

> Would it be possible to accept a list of domains separated by commas?
eg. "myapp.internal" or ["foo.internal", "bar.internal"] or "foo.internal,bar.internal"

It is certainly possible, although I don't recommend introducing this to the API of this library. While not explicitly mentioned in the [anti-patterns](https://hexdocs.pm/elixir/1.16.0-rc.0/design-anti-patterns.html#using-application-configuration-for-libraries) guides, I believe it is trivial to do the string splitting in your application code. e.g.

```elixir
# config/runtime.exs
config :my_app, dns_cluster_query: String.split(System.get_env("DNS_CLUSTER_QUERY"), ",")

# lib/application.ex
{DNSCluster, query: Application.get_env(:my_app, dns_cluster_query)}
```

I think this keeps the library API simple while enabling maximum flexibility for application developers using this library

davydog187 commented:
Also, I suppose that it is valid to have a `,` in your DNS query, so that's another reason not to support it

davydog187 committed:
Support differing basenames

davydog187 commented:
@josevalim @chrismccord I have updated this PR to add support for connecting to nodes with a basename different than the host. To support this, I've extended the API to allow the user to specify a query, a `{basename, query}` or a list of either.

I've updated the documentation to reflect this change, please let me know if you want to go a different direction. I'm not an expert in erlang distribution, but I'm not aware of another parameter that would warrant a more complicated datastructure. If that is the case, we should probably use a map or struct instead.

davydog187 committed:
edit docs for clarity

davydog187 committed:
revert first sentence

davydog187 commented:
I'm running this branch in production and it works great!

<img width="392" alt="Screenshot 2024-01-13 at 9 32 34 AM" src="https://github.com/phoenixframework/dns_cluster/assets/1019721/96a69808-f9b7-4111-a83e-49f0f31a525c">

davydog187 commented:
I'm running this branch in production and it works great! (Redacted IPs)

```elixir
iex> Node.list()
[
    :"sauron-prod@fdaa:1:aaaaaaaaaaaaaaaaaaaaa:2",
    :"leaderboard-prod@fdaa:1:aaaaaaaaaaaaaaaaaaaaaa:2",
    :"leaderboard-prod@fdaa:1:aaaaaaaaaaaaaaaaaaaaaa:2"
]
```

davydog187 commented on lib/dns_cluster.ex:
```diff
@@ -151,13 +165,39 @@ defmodule DNSCluster do
+  defp valid_query?({basename, query}) when is_binary(basename) and is_binary(query), do: true
+
+  defp valid_query?(list) when is_list(list) do
+    Enum.all?(list, &valid_query?/1)
```
This allows nested lists which is not intended

davydog187 committed:
don't allow nested lists for queries

davydog187 commented:
For anyone interested in this feature, I encourage you to test it out! I've been running it in production for two weeks now without any issue

schrockwell commented:
@davydog187 Thanks, I'm trying this on Fly right now. To set it up, I had to do a couple things:

1. In `rel/env.sh.eex`, set `export RELEASE_NODE="${FLY_APP_NAME}@${FLY_PRIVATE_IP}"` (note that `$FLY_IMAGE_REF` has been removed from the node basename)
2. Parse the `$DNS_CLUSTER_QUERIES` which is in the format `"app1@app1.internal app2@app2.internal"`:
```elixir
dns_cluster_queries =
  System.get_env("DNS_CLUSTER_QUERIES", "")
  |> String.split(" ", trim: true)
  |> Enum.map(fn query ->
    case String.split(query, "@") do
      [basename, domain] -> {basename, domain}
      [domain] -> domain
    end
  end)
```
3. Set `DNS_CLUSTER_QUERIES` in both apps' TOML configs and deploy.

davydog187 commented:
Hey @schrockwell 👋🏼 

That all sounds well and good, however, I would personally just hardcode the queries in your TOML file, you shouldn't need to parse them given that they should be static per environment

davydog187 commented:
Sorry to ping you @chrismccord but is there anything holding up this PR besides being a low priority?

yordis commented:
Hey folks, any opportunity to get this one in? 🙏🏻

chrismccord commented:
❤️❤️❤️🐥🔥
